File: Course_2-Data_Structures/Week-1/Excercise_Challenges/4_stack_with_max/stack_with_max_fast.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include <algorithm>
using namespace std;

class StackWithMax{
    public:
        vector<int> stack;
        vector<int> max_stack;

        void Push(int value){
            stack.push_back(value);
            
            if(max_stack.size() == 0)
                max_stack.push_back(value);
            else
                max_stack.push_back(max(max_stack.back(), value));
        }

        void Pop(){
            assert(stack.size());
            stack.pop_back();
            max_stack.pop_back();
        }

        int Max(){
            assert(stack.size());
            return max_stack.back();
        }
};

int main(void)
{
    int num_queries;
    cin >> num_queries;

    StackWithMax stack;
    string query;
    for(int i = 0; i < num_queries; ++i){
        cin >> query;

        if(query == "push"){
            int value;
            cin >> value;
            stack.Push(value);
        }
        else if(query == "pop"){
            stack.Pop();
        }
        else if(query == "max"){
            cout << stack.Max() << "\n";
        }
        else{
            assert(0);
        }
    }
    
    return 0;
}

File: Course_2-Data_Structures/Week-1/Excercise_Challenges/4_stack_with_max/stack_with_max_fast.py
class StackWithMax():
    def __init__(self):
        self.stack = []
        self.max_stack = []

    def Push(self, a):
        self.stack.append(a)

        if len(self.max_stack) == 0:
            self.max_stack.append(a)
        else:
            self.max_stack.append(max(self.max_stack[-1], a))

    def Pop(self):
        assert(len(self.stack))
        self.stack.pop()
        self.max_stack.pop()

    def Max(self):
        assert(len(self.max_stack))
        return self.max_stack[-1]


if __name__ == '__main__':
    stack = StackWithMax()

    num_queries = int(input())
    for _ in range(num_queries):
        query = input().split()

        if query[0] == "push":
            stack.Push(int(query[1]))
        elif query[0] == "pop":
            stack.Pop()
        elif query[0] == "max":
            print(stack.Max())
        else:
            assert(0)
